docs/proposals/20230613-node-prediction.md

```diff
@@ -0,0 +1,407 @@
+    // MidMemoryThresholdPercent defines the maximum percentage of the Mid-tier memory resource dividing the node allocatable.
+    // MidMemoryAllocatable <= NodeMemoryAllocatable * MidMemoryThresholdPercent / 100.
+    MidMemoryThresholdPercent *int64 `json:"midMemoryThresholdPercent,omitempty" validate:"omitempty,min=0,max=100"`
+}
+```
+
+##### Extended Resources
+
+```yaml
+apiVersion: v1
+kind: Node
+metadata:
+  name: test-node
+status:
+  allocatable:
+    cpu: '32'
+    memory: 129636240Ki
+    pods: '213'
+    kubernetes.io/mid-cpu: '16000' # allocatable cpu milli-cores for Mid-tier pods
+    kubernetes.io/mid-memory: 64818120Ki # allocatable memory bytes for Mid-tier pods
+  capacity:
+    cpu: '32'
+    memory: 129636240Ki
+    pods: '213'
+    kubernetes.io/mid-cpu: '16000'
+    kubernetes.io/mid-memory: 64818120Ki
+```
+
+### Theoretical Model
+
+#### Node Peak Prediction
+
+Before elaborating the peak prediction algorithm, let's formalize the node peak prediction problem.
+
+Let's denote the usage of a Pod `p` at the time `t` is `U(p, t)`.
+
+Then the usage of a Node `M` which schedules a set of Pods is `MU(Pods, t) = sum[p in Pods](U(p, t))`.
+
+> Note that the non-Pod usage of the node can be regarded as the usage of a special pod `S`.
+
+When we want to predict the node peak at the time `T`, we are calculating
+`Peak(Pods, T) = max[t >= T](sum[p in Pods](U(p, t)))`.
+
+The predicted peak `Peak(Pods, T)` is our node prediction result at `T`.
+
+#### N-sigma Prediction
+
+There are several [statistical peak prediction models](#alternatives) which are practical to implement in the online
+scheduler. [*N-sigma*](#references) is the picked peak prediction model in the current implementation. It assumes the
+timing node metrics follow the Gaussian distribution, which allows us to estimate the node peak with the mean and
+standard deviation (stdev):
+
+`Peak_N-Sigma_v1(Pods, T) = mean[T0 <= t <= T](MU(Pods, t)) + N * stdev[T0 <= t <= T](MU(Pods, t))`
+
+The `Peak_N-Sigma_v1` is the predicted node peak. It is implemented as the first version of node prediction, which is
+calculated based on node-level metrics.
+
+Moreover, we can calculate with the pods' metrics:
+
+`Peak_Pods-N-Sigma'(Pods, T) = sum[p in Pods](mean[T0 <= t <= T](U(p, t)) + N * stdev[T0 <= t <= T](U(p, t)))`
+
+A more conservative is derived from their maximal. The `Peak_N-sigma_v2` is the second version of node prediction,
+which also considers the pod-level metrics.
+
+`Peak_N-Sigma_v2(Pods, T) = max(Peak_N-Sigma_v1(Pods, T), Peak_Pods-N-Sigma(Pods, T))`.
+
+#### Mid-tier Overcommitment
+
+The Mid-tier resource contains the reclaimable resources which are probably unused in the long-term by the
+high-priority (i.e. Prod) pods.
+
+The resource calculation for the Mid-tier resources can be described as follows:
+
+```
+Allocatable[Mid] := min(Reclaimable[Mid], NodeAllocatable * thresholdRatio)
```

Comment:
Thinking about the ledger needs to add a description. Currently, only reclaimable resources are visible in the mid band, which is a temporary solution. In the future, we may hope to merge the prod + mid ledger to obtain better flexibility, but the corresponding price is prod when resources are insufficient, it will preempt the resources of mid.

Reply:
added